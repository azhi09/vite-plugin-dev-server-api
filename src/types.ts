import { FilterPattern } from "vite";

export type ApiFunction = () => ApiReturnBase | Promise<() => ApiReturnBase>;

export type ApiReturnBase = object | string | number | boolean;
export type ApiReturn = ApiFunction | ApiReturnBase;

export type Api = {
  [url: string]: ApiReturn;
};

export type ApiMapUrlValue = {
  type: "json" | "object" | "function";
  path?: string;
  code?: ApiReturn;
};

/** url与文件的对应关系 */
export type ApiMapUrl = {
  [url: string]: ApiMapUrlValue;
};

export type PkgType = "commonjs" | "module";

export interface RawOptions {
  include?: FilterPattern;
  exclude?: FilterPattern;
  api?: Api | (() => Api) | (() => Promise<Api>);
  dir?: string;
  tempDir?: string;
  cwd?: string;
  fileFormat?: string;
  fileFormatExcludeJson?: string[];
}

export type Options = Omit<
  Required<RawOptions>,
  "dir" | "tempDir" | "cwd" | "fileFormat" | "fileFormatExcludeJson"
> & {
  dir: string;
  dirPath: string;
  tempDir: string;
  tempDirPath: string;
  cwd: string;
  fileFormat: string;
  fileFormatExcludeJson: string[];
};
