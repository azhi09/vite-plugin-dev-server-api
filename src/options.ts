import { Options, RawOptions } from "./types";
import path from "path";

const defaultOptions = {
  dir: "api",
  tempDir: ".temp",
  cwd: process.cwd(),
  /** api file format */
  fileFormat: "cjs,mjs,js,ts,json",
  fileFormatExcludeJson: [".cjs", ".mjs", ".js", ".ts"],
};

function setFileFormatExcludeJson(fileFormat: string): string[] {
  return fileFormat
    .split(",")
    .reduce((previousValue: string[], currentValue: string) => {
      return currentValue === "json"
        ? [...previousValue]
        : [...previousValue, `.${currentValue}`];
    }, []);
}

export function setOptions(argv: RawOptions, root: string): Options {
  const options = Object.assign({}, defaultOptions, argv) as Options;
  options.dirPath = path.join(root, options.dir);
  options.tempDirPath = path.join(
    root,
    "node_modules",
    "vite-plugin-dev-server-api",
    options.tempDir
  );
  options.fileFormatExcludeJson = setFileFormatExcludeJson(options.fileFormat);

  return options;
}
