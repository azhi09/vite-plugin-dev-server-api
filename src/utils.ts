import { Api, ApiMapUrl, RawOptions } from "./types";
import fs from "fs/promises";
import path from "path";
import { transformWithEsbuild } from "vite";

export function isFunction(fn: unknown): fn is () => unknown {
  return typeof fn === "function";
}

export function resolveJsonFile(code: string, filePath: string): ApiMapUrl {
  let ctn;
  try {
    ctn = JSON.parse(code);
  } catch (e) {
    console.error(`${filePath} parse error.`);
    ctn = {};
  }

  return reduceJson(ctn, filePath);
}

export function reduceJson(ctn: object, filePath: string): ApiMapUrl {
  return Object.keys(ctn).reduce(
    (previousValue: ApiMapUrl, currentValue: string) => {
      const res = previousValue;
      res[currentValue] = {
        type: "json",
        path: filePath,
      };
      return res;
    },
    {}
  );
}

export async function resolveTsFile(
  code: string,
  filePath: string,
  tempDirPath: string,
  isEsm: boolean
): Promise<ApiMapUrl> {
  const { code: transformCodeResultCode } = await transformCode(
    code,
    filePath,
    isEsm
  );
  let res = {};
  if (transformCodeResultCode) {
    const transformCodeResultFilePath = path.join(
      tempDirPath,
      `${Date.now()}.${isEsm ? "mjs" : "js"}`
    );

    await fs.writeFile(transformCodeResultFilePath, transformCodeResultCode);

    const fileCode = await import(transformCodeResultFilePath),
      fileCodeIsFunction = isFunction(fileCode.default);

    let ctn = {};

    if (fileCode.default) {
      if (fileCodeIsFunction) {
        ctn = await fileCode.default();
      } else {
        ctn = fileCode.default;
      }
    } else {
      console.error(`${filePath} default module not found.`);
    }
    res = reduceTs(ctn, transformCodeResultFilePath, fileCodeIsFunction);
  }

  return res;
}

export async function transformCode(
  code: string,
  filePath: string,
  isEsm: boolean
): Promise<{ code: string }> {
  let result;
  try {
    result = await transformWithEsbuild(code, filePath, {
      format: isEsm ? "esm" : "cjs",
    });
  } catch {
    console.error(`${filePath} parse error`);
    result = { code: "" };
  }

  return result;
}

export async function reduceTs(ctn: object, filePath: string, isFn: boolean) {
  return Object.keys(ctn).reduce(
    (previousValue: ApiMapUrl, currentValue: string) => {
      const res = previousValue;
      res[currentValue] = {
        type: isFn ? "function" : "object",
        path: filePath,
      };
      return res;
    },
    {}
  );
}

export async function resolveApi(
  api: Required<RawOptions["api"]>
): Promise<ApiMapUrl> {
  let ctn: Api = {};
  if (isFunction(api)) {
    ctn = (await api()) as Api;
  } else {
    ctn = api as Api;
  }
  return ctn ? reduceApi(ctn) : {};
}

export function reduceApi(ctn: Api): ApiMapUrl {
  return Object.keys(ctn).reduce(
    (previousValue: ApiMapUrl, currentValue: string) => {
      const res = previousValue;
      res[currentValue] = {
        type: isFunction(ctn[currentValue]) ? "function" : "object",
        code: ctn[currentValue],
      };

      return res;
    },
    {}
  );
}
