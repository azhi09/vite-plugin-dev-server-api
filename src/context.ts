import { setOptions } from "./options";
import {
  ApiMapUrl,
  ApiMapUrlValue,
  ApiReturnBase,
  Options,
  RawOptions,
  PkgType,
} from "./types";
import path from "path";
import fs from "fs/promises";
import { rimraf } from "rimraf";
import FastGlob from "fast-glob";
import {
  isFunction,
  resolveApi,
  resolveJsonFile,
  resolveTsFile,
} from "./utils";

export class Context {
  options: Options;
  root = process.cwd();

  constructor(private rawOptions: RawOptions = {}) {
    const { cwd } = rawOptions;

    if (cwd) {
      this.root = cwd;
    }

    this.options = setOptions(rawOptions, this.root);
  }

  async setTempDir() {
    const { tempDirPath } = this.options;

    try {
      await fs.access(tempDirPath);
      await rimraf(tempDirPath);
      await fs.mkdir(tempDirPath);
    } catch (e) {
      await fs.mkdir(tempDirPath);
    }
  }

  async isEsm() {
    const pkg = await fs.readFile(path.join(this.root, "package.json"), {
        encoding: "utf-8",
      }),
      pkgType: PkgType = JSON.parse(pkg).type || "commonjs";
    return "module" === pkgType;
  }

  async findApiMap(): Promise<ApiMapUrl> {
    const filePaths: string[] = await FastGlob(
      `./**/*.{${this.options.fileFormat}}`,
      {
        cwd: this.options.dirPath,
        absolute: true,
      }
    );
    const isEsm = await this.isEsm();
    let res = {};

    for (const filePath of filePaths) {
      const extname = path.extname(filePath),
        code = await fs.readFile(filePath, { encoding: "utf-8" });

      switch (extname) {
        case ".json":
          res = { ...res, ...resolveJsonFile(code, filePath) };
          break;

        case ".js":
        case ".cjs":
        case ".mjs":
        case ".ts":
          res = {
            ...res,
            ...(await resolveTsFile(
              code,
              filePath,
              this.options.tempDirPath,
              isEsm
            )),
          };
          break;

        default:
          break;
      }
    }

    return res;
  }

  async apiToApiMap(): Promise<ApiMapUrl> {
    return await resolveApi(this.options.api);
  }

  async getApiMapUrl(): Promise<ApiMapUrl> {
    await this.setTempDir();

    return {
      ...(await this.findApiMap()),
      ...(await this.apiToApiMap()),
    };
  }

  async getApiResult(
    url: string,
    apiMapUrlValue: ApiMapUrlValue
  ): Promise<ApiReturnBase> {
    const { type: apiType, path, code } = apiMapUrlValue;
    const isFn = apiType === "function";

    let content, parseContent, result;

    /** 文件 / code */
    if (code) {
      content = code;
    } else {
      const importCtn = await import(path as string);

      content = importCtn.default;
    }

    if (isFn) {
      parseContent = await content();
    } else {
      parseContent = content;
    }

    const urlContent = parseContent[url];

    if (isFunction(urlContent)) {
      result = await urlContent();
    } else {
      result = urlContent;
    }

    return result;
  }
}
