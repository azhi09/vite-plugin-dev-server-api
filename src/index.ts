import type { Plugin, ViteDevServer } from "vite";
// import DEBUG from "debug";
import { ApiMapUrl, ApiReturnBase, RawOptions } from "./types";
import { Context } from "./context";

async function devServerApi(argv: RawOptions = {}): Promise<Plugin> {
  const ctx = new Context(argv);

  let apiMapUrl: ApiMapUrl = await ctx.getApiMapUrl();

  return {
    name: "vite-plugin-dev-server-api",

    configureServer(server: ViteDevServer) {
      server.middlewares.use(async (req, res, next) => {
        const { url } = req;
        let result: ApiReturnBase = {
          message: `[${url}] not found.`,
        };

        if (url && url in apiMapUrl) {
          result = await ctx.getApiResult(url, apiMapUrl[url]);

          res.end(JSON.stringify(result));
        } else {
          next();
        }
      });

      server.watcher.on("all", async (evt, path) => {
        const { dirPath } = ctx.options;

        if (path.startsWith(dirPath)) {

          apiMapUrl = await ctx.getApiMapUrl();
        }
      });
    },
  };
}

export default devServerApi;
