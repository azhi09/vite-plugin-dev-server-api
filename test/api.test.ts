import { describe, expect, it } from "vitest";
import { Context } from "../src/context";
import path from "path";

const api = {
  "/api/test/object": { data: "object" },
  "/api/test/function": () => ({
    data: "object",
  }),
  "api/test/promiseFunction": () =>
    new Promise((res) => {
      setTimeout(() => {
        res({ data: "Promise<object>" });
      }, 3 * 1000);
    }),
};

describe("api", () => {
  it("find api", async () => {
    const cwd = path.join(process.cwd(), "examples/vite-vue/");

    const ctx = new Context({ cwd });

    const apiMapUrl = await ctx.findApiMap();

    // expect(apiMapUrl).toMatchSnapshot();
  });

  it("api to apiMap", async () => {
    const ctx = new Context({ api });
    const ctxApiMapUrl = await ctx.apiToApiMap();

    expect(ctxApiMapUrl).toMatchSnapshot();
  });
});
