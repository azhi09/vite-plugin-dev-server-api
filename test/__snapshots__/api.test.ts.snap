// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`api > api to apiMap 1`] = `
{
  "/api/test/function": {
    "code": [Function],
    "type": "function",
  },
  "/api/test/object": {
    "code": {
      "data": "object",
    },
    "type": "object",
  },
  "api/test/promiseFunction": {
    "code": [Function],
    "type": "function",
  },
}
`;
